src/Uno.UI/UI/Xaml/Input/PointerRoutedEventArgs.Android.cs

```diff
@@ -1,37 +1,156 @@
+			var updates = _none;
+			switch (type)
+			{
+				case MotionEventToolType.Finger:
+					props.IsLeftButtonPressed = Pointer.IsInContact;
+					updates = isDown ? _fingerDownUpdates : isUp ? _fingerUpUpdates : _none;
+					break;
+				case MotionEventToolType.Mouse:
+					props.IsLeftButtonPressed = _nativeEvent.IsButtonPressed(MotionEventButtonState.Primary);
+					props.IsMiddleButtonPressed = _nativeEvent.IsButtonPressed(MotionEventButtonState.Tertiary);
+					props.IsRightButtonPressed = _nativeEvent.IsButtonPressed(MotionEventButtonState.Secondary);
+					updates = isDown ? _mouseDownUpdates : isUp ? _mouseUpUpdates : _none;
+					break;
+				case MotionEventToolType.Stylus:
+					props.IsBarrelButtonPressed = _nativeEvent.IsButtonPressed(MotionEventButtonState.StylusPrimary);
+					props.IsLeftButtonPressed = !props.IsBarrelButtonPressed;
+					break;
+				case MotionEventToolType.Eraser:
+					props.IsEraser = true;
+					break;
+				case MotionEventToolType.Unknown: // used by Xamarin.UITest
+				default:
+					break;
+			}
+
+			if (updates.TryGetValue(_nativeEvent.ActionButton, out var update))
+			{
+				props.PointerUpdateKind = update;
+			}
+
+			return props;
+		}
+
+		#region Misc static helpers
+		private static readonly Dictionary<MotionEventButtonState, PointerUpdateKind> _none = new Dictionary<MotionEventButtonState, PointerUpdateKind>(0);
+		private static readonly Dictionary<MotionEventButtonState, PointerUpdateKind> _fingerDownUpdates = new Dictionary<MotionEventButtonState, PointerUpdateKind>
+		{
+			{ MotionEventButtonState.Primary, PointerUpdateKind.LeftButtonPressed }
+		};
+		private static readonly Dictionary<MotionEventButtonState, PointerUpdateKind> _fingerUpUpdates = new Dictionary<MotionEventButtonState, PointerUpdateKind>
+		{
+			{ MotionEventButtonState.Primary, PointerUpdateKind.LeftButtonReleased }
+		};
+		private static readonly Dictionary<MotionEventButtonState, PointerUpdateKind> _mouseDownUpdates = new Dictionary<MotionEventButtonState, PointerUpdateKind>
+		{
+			{ MotionEventButtonState.Primary, PointerUpdateKind.LeftButtonPressed },
+			{ MotionEventButtonState.Tertiary, PointerUpdateKind.MiddleButtonPressed },
+			{ MotionEventButtonState.Secondary, PointerUpdateKind.RightButtonPressed }
+		};
+		private static readonly Dictionary<MotionEventButtonState, PointerUpdateKind> _mouseUpUpdates = new Dictionary<MotionEventButtonState, PointerUpdateKind>
+		{
+			{ MotionEventButtonState.Primary, PointerUpdateKind.LeftButtonReleased },
+			{ MotionEventButtonState.Tertiary, PointerUpdateKind.MiddleButtonReleased },
+			{ MotionEventButtonState.Secondary, PointerUpdateKind.RightButtonReleased }
+		};
+
+		private static readonly ulong _unixEpochMs = (ulong)(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc) - new DateTime()).TotalMilliseconds;
+
+		private static ulong ToTimeStamp(long uptimeMillis)
+		{
+			// We cannot cache the "bootTime" as the "uptimeMillis" is frozen while in deep sleep
```

Comment:
Can't you use `Stopwatch` instead ? It'll be less costly than going through the JNI.

Reply:
The issue is that here I convert a timestamp provided by the system in a kind of random scale, into an absolute value. I don't track the interval by myself, and I cannot as I may have to convert the same timestamp multiple times (if the event is bubbling natively), result would not be predictable. I don't see how I could do that using a `StopWatch` (and actually I would think that it will also access to the system time ...)

However, I agree that the resulted timestamps are mainly used only to compute interval between some pointer events in a short time range. So there is no real needs to get this timestamp in an absolute scale. I'll add a feature flag that avoids this calculation (enabled by default).